Issue 5659 - local fonts on editor

myroslavve opened the pull request:
# Description
This update makes loading fonts slower for local fonts, because we have to wait for the font to load from google/bunny to local server and then wait for fonts to load from server to editor. There is nothing we can do about it, maybe we should add spinning wheel to the font selector to show when the font is loading now.
<!---
Please include a summary of the changes and the related issue.
Please also include relevant motivation and context.
--->

Fixes #5659 

# How Has This Been Tested?
Enabled local fonts in the dashboard, on editor loaded any font, checked in console that path to that font is "localhost...".
<!---
Please describe the tests that you ran to verify your changes.
Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration.
--->

# Test checklist

<!--- Please remove the unnecessary checkbox --->

**_ Front/Back Testing _**

-   [ ] Test the component in normal state in sidebar
-   [ ] Test the component in hover state in sidebar (if hover exists)
-   [ ] Check that the settings work on frontend
-   [ ] Check that backend works and saves the settings after the editor reload
-   [ ] Same 1-4 points for the toolbar
-   [ ] Same 1-4 points on responsive
-   [ ] Test the block inside the grid (Container + Row + Column)
-   [ ] Test the block as a standalone block
-   [ ] Duplicate the block, test that the settings of the first do not affect the second
-   [ ] Test with 2 blocks of the same type

**_ Pre-Code Testing _**

-   [ ] Test the component in normal state in sidebar
-   [ ] Test the component in hover state in sidebar (if hover exists)
-   [ ] Check that the settings work on frontend
-   [ ] Check that backend works and saves the settings after the editor reload
-   [ ] Same 1-4 points for the toolbar
-   [ ] Same 1-4 points on responsive
-   [ ] Test the block inside the grid (Container + Row + Column)
-   [ ] Test the block as a standalone block
-   [ ] Duplicate the block, test that the settings of the first do not affect the second
-   [ ] Test with 2 blocks of the same type
-   [ ] Check no commented code and no unnecessary imports
-   [ ] Standards of the project have been followed
-   [ ] No errors/warnings on console

# Checklist

-   [ ] My code follows the style guidelines of this project
-   [ ] I have performed a self-review of my code
-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] My changes generate no new warnings/errors
-   [ ] I have added/updated tests that prove my fix is effective or that my feature works
-   [ ] New and existing unit tests pass locally with my changes


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a new REST API endpoint to dynamically retrieve font URLs.
  - Enhanced loading state management across various components for a better user experience during font operations.
  - Added an `overlay` prop to the ContentLoader component for dynamic styling capabilities.

- **Bug Fixes**
  - Improved error handling and clarity in font CSS file uploads.

- **Refactor**
  - Simplified font URL generation logic for improved maintainability.
  - Modularized font URL construction and loading logic for better readability and clarity.
  - Transitioned to a singleton pattern for font management.
  - Improved naming conventions for methods and properties to enhance code clarity.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

myroslavve committed:
Create api endpoint which returns font url markup

myroslavve committed:
Update store to handle requests for the font url

myroslavve committed:
Refactor and update `getFontUrl` to use local fonts on editor

myroslavve committed:
Optimize slightly by skipping `fontDataString` generation for local fonts

myroslavve committed:
Make local fonts work on editor

myroslavve committed:
Refactor `get_maxi_blocks_font_url`

elzadj reviewed (changes requested):
Getting weird results after I re-open the editor:

[screencast-bpconcjcammlapcogcnnelfmaeghhagj-2024.07.26-15_32_59 (1).webm](https://github.com/user-attachments/assets/41a82b5c-b403-426e-8e92-3bed6f762d2a)

Tests tell us something is breaking, too.

Great idea about a loader, but I would add it to the Text (Button, etc) block itself, maybe with "Loading font..." text, but I guess the one near the Font selector is easier to do.

elzadj commented:
@Fybex test this one too please, when you have time. Thanks!

elzadj commented:
@myroslavve @Olekrut can't recreate the bugs from my video, btw, so could be something local.
But the tests are failing hard, so something is not right, needs checking.

Fybex commented on src/extensions/text/store/resolvers.js:
```diff
@@ -0,0 +1,21 @@
+	getFontUrl:
+		(fontName, fontData) =>
+		async ({ dispatch }) => {
+			const fontUrl = JSON.parse(
```
Looks like JSON.parse is unnecessary if avoid double encoding by removing wrapping url in [class-maxi-api](https://github.com/maxi-blocks/maxi-blocks/blob/b66b98a04a5a02e301fb8ece48e692239f372747/core/class-maxi-api.php#L704)

myroslavve replied:
sure, this works :+1:

myroslavve committed:
Fix error shown in tests

myroslavve committed:
Rename methods and variables of local fonts to snake case

myroslavve committed:
Refactor local fonts abstract sanitize font name

myroslavve committed:
Remove encoding for font url

myroslavve committed:
Merge branch 'master' into issue-5659

myroslavve committed:
Fix styles class using old function name

myroslavve commented:
@Olekrut test with bunny fonts again pls

Olekrut reviewed (approved):
![image](https://github.com/user-attachments/assets/63311ffd-28b3-4ec5-9446-87df74ec026e)

elzadj reviewed (changes requested):
Almost, almost, great work so far! 

1. I still see a space in the id on backend if the font name has 2 and more spaces. Need to replace globally.
![Screenshot from 2024-08-02 16-22-49](https://github.com/user-attachments/assets/a87462b0-0676-455c-9c80-10f9e1e5117e)

2. There is a several seconds visual delay when we change a font while it's uploading, we really need some sort of indicator here. @ckp267 @kyrapieterse ? I think we should add a loading animation to the whole text blocks, maybe even re-use our usual loader, but with some semi-opacity for the block.

[screencast-wordpress.local-2024.08.02-16_21_14.webm](https://github.com/user-attachments/assets/366a0e6b-4c02-4d9f-9450-4c9d551bbaf8)

ckp267 commented:
> 2. There is a several seconds visual delay when we change a font while it's uploading, we really need some sort of indicator here. @ckp267 @kyrapieterse ? I think we should add a loading animation to the whole text blocks, maybe even re-use our usual loader, but with some semi-opacity for the block.
> 
>  screencast-wordpress.local-2024.08.02-16_21_14.webm

As per @elzadj 

**My idea for the loading - make the text 0.5 opacity and show our usual loader circle.**

---
Note from Llama 405B as a summary of internal discussion thread:
- The delay in displaying a font is due to the need to download it first. To address this, a loading indicator will be implemented, which will display a loader circle and reduce the text opacity to 0.5. 
- The idea of previewing the font style in the dropdown was discussed, but it was deemed impractical due to the need to load all fonts or create images for each font. 
- Instead, the focus will be on implementing a loader that indicates the font is being downloaded. Unused fonts will be removed periodically, specifically each time all fonts are loaded, to prevent accumulation of unnecessary files. 
- The loader will not include any text, such as "Downloading fonts...", due to potential layout and language translation issues.

myroslavve committed:
Replace all spaces in font id

myroslavve committed:
Add loader over existing block in MaxiBlock

myroslavve committed:
Implement loading animation on font load

myroslavve committed:
Make block transparent while loading font

elzadj reviewed (changes requested):
Works great! Now need to add it to:
1. Image caption text font
2. Number Counter number
3. Accordion 
4. Search input


@Olekrut did I miss something?

Olekrut reviewed (changes requested):
Then need to do the same for SC as well
and for Map block

About loader:
Looks like with loader is longer than on master
this branch:
https://jam.dev/c/3329405e-b3c9-4c53-aeb5-80fe453dded5
master:
https://jam.dev/c/57e79f36-9863-44c0-8b0d-d867e5aded8c

myroslavve committed:
Fix loading animation interrupting when selecting another font

myroslavve committed:
Add loader to some of the missed blocks

myroslavve committed:
Make loader always appear on top

Was not appearing for search maxi

myroslavve committed:
Add loader to map maxi

elzadj commented:
@myroslavve great! 
Toolbar needs it too:
![screenshot-wordpress local-2024 08 09-12_56_35](https://github.com/user-attachments/assets/bab90df4-e7b0-4bc5-812c-f41bea9e2278)

And SC, but not sure where to add the loader here - main thing for it to be visible.

myroslavve committed:
Add loader to toolbar and caption toolbar

myroslavve committed:
Add loader to style cards font selectors

myroslavve committed:
Merge branch 'master' into issue-5659

myroslavve committed:
Fix infinite loading on SC dropdowns

myroslavve commented:
> Then need to do the same for SC as well and for Map block

Added.

> About loader: Looks like with loader is longer than on master this branch: https://jam.dev/c/3329405e-b3c9-4c53-aeb5-80fe453dded5 master: https://jam.dev/c/57e79f36-9863-44c0-8b0d-d867e5aded8c

The reason we added the loader is that other changes in this PR made loading fonts slower.

elzadj reviewed (changes requested):
Niice, found that for H1 in a custom SC the loader gets stuck (and then all other loaders after it, too). I've used an old pattern from the master.
![screenshot-wordpress local-2024 08 20-13_01_54](https://github.com/user-attachments/assets/b567cd8d-e1d9-4ddd-b8a6-22bdf0c34875)

myroslavve committed:
use singleton for `MaxiBlocks_Local_Fonts` everywhere (refactoring)

myroslavve commented:
> Niice, found that for H1 in a custom SC the loader gets stuck (and then all other loaders after it, too). I've used an old pattern from the master. ![screenshot-wordpress local-2024 08 20-13_01_54](https://private-user-images.githubusercontent.com/2401618/359436061-b567cd8d-e1d9-4ddd-b8a6-22bdf0c34875.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNDk4ODEsIm5iZiI6MTcyNDE0OTU4MSwicGF0aCI6Ii8yNDAxNjE4LzM1OTQzNjA2MS1iNTY3Y2Q4ZC1lMWQ5LTRkZGQtYjhhNi0yMmJkZjBjMzQ4NzUuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjBUMTAyNjIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDcxMzRkZGEwOWU2MTlhNjQwZTgwOTE5ZTdhMDYyZWIxY2FlMmZlMDFmZWFhNTA5MTIzOTNiMzFmNWU0NDQ2MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.VpoLfCzxBW-J3kJ5RJoajTjkoO4bNWnzeSqZ5e7Lbhk)

Do you have any errors in console? I fixed this issue yesterday and cannot reproduce it on the latest version of this branch.

elzadj commented:
> Do you have any errors in console? I fixed this issue yesterday and cannot reproduce it on the latest version of this branch.

Not really:

[screencast-bpconcjcammlapcogcnnelfmaeghhagj-2024.08.20-17_47_27.webm](https://github.com/user-attachments/assets/7c33a0ef-cdae-4c7e-ae44-8ef8ad116c9c)

Happens both with google and bunny fonts

elzadj commented:
@Olekrut Do you see the same bug? ^

Olekrut commented:
> Do you see the same bug? ^

Actually, i can re-create it everywhere ( P, button, H1-6 ) if change font again before it loads.
Everything is gets ok if to wait 2-3 min, but yes, it is not ok

elzadj commented:
@myroslavve what's the status of this one? Maybe let's remove the loader from the SC for now, if it's easier to do? Then we can check if we have any support requests about it.

kyrapieterse commented:
> @myroslavve what's the status of this one? Maybe let's remove the loader from the SC for now, if it's easier to do? Then we can check if we have any support requests about it.

Agree = remove what we dont need

elzadj commented:
> > @myroslavve what's the status of this one? Maybe let's remove the loader from the SC for now, if it's easier to do? Then we can check if we have any support requests about it.
> 
> I still cannot consistently reproduce this issue, so let's do it separately, because it can take some time to figure out.

Good idea, but remove the loader from the SC for now, please, in case we need to update the plugin before it's fixed. Thanks!

myroslavve committed:
Revert "Add loader to style cards font selectors" and "Fix infinite loading on SC dropdowns"

This reverts commits 08c41852b5e5563dae3fd3f373b1f695a910ded2
d254c76a3b03a9f0e3f461f8b3b641f9113ec71d

elzadj commented:
I merge it as soon as the tests pass. @myroslavve create a new issue for the SC loader, please.

elzadj reviewed (changes requested):
Something's wrong with frontend when I use bunny local

[screencast-wordpress.local-2024.08.27-18_00_39 (1).webm](https://github.com/user-attachments/assets/42ced772-bff8-48b3-ab17-6d032f864010)

Olekrut reviewed (changes requested):
the same with google fonts, but works only custom  fonts and only on frontend
![image](https://github.com/user-attachments/assets/6ac6ba61-1a25-400c-a481-61d1ef59b140)

myroslavve committed:
Fix style cards fonts not loading on frontend

myroslavve committed:
Merge branch 'master' into issue-5659

elzadj commented:
@myroslavve not sure how it can affect the backgrounds, but the tests clearly show it :melting_face:

Olekrut reviewed (changes requested):
![image](https://github.com/user-attachments/assets/f9727df7-ea67-4d0e-a014-da420f4e3dae)
frontend

myroslavve committed:
Fix frontend error for fonts

myroslavve commented:
> ![image](https://private-user-images.githubusercontent.com/88893350/363663774-f9727df7-ea67-4d0e-a014-da420f4e3dae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUzMTE4MzksIm5iZiI6MTcyNTMxMTUzOSwicGF0aCI6Ii84ODg5MzM1MC8zNjM2NjM3NzQtZjk3MjdkZjctZWE2Ny00ZDBlLWEwMTQtZGE0MjBmNGUzZGFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAyVDIxMTIxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiY2YyZDlhZTE2YjAwZTU2ZmQ3ZDIwOTAyYWMwOTlmYTVkYzYxOWNlMTk1MDQzODM5Y2RjYjYxM2ViZjVmOGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TA9gQMU42D41xckuK5eWPfzWrIS3EWzmgxSFQAhhi1A) frontend

fixed

Olekrut reviewed (changes requested):
https://jam.dev/c/df9fc381-e949-455a-bcdb-2b9849729aa1
local fonts are loading very long (3-4min) on backend
https://jam.dev/c/f03c5d85-491f-40e4-b108-f7ca523fb308

myroslavve commented:
> https://jam.dev/c/df9fc381-e949-455a-bcdb-2b9849729aa1 local fonts are loading very long (3-4min) on backend https://jam.dev/c/f03c5d85-491f-40e4-b108-f7ca523fb308

Tried to optimize, but everything is already as fast as it can be for local fonts, at least logically. With local fonts the speed depends a lot on the speed of the server, so on local it will be slower than on an actual server where WordPress would be.
